infor-design/enterprise-wc
# 153 - IdsDataGrid/IdsZip excel export feature

**Explain the details for making this change. What existing problem does the pull request solve?**
Adds feature to export data to csv and xlsx excel files.
.xlsx are zip files, so an IdsZip util module was also added

Relevant doc links are provided in class/function comments, two important ones listed below
Office open xml docs - https://www.ecma-international.org/publications-and-standards/standards/ecma-376/ (Part 1 pdf)
Zip file format docs - https://docs.fileformat.com/compression/zip/

**Related github/jira issue (required)**:
Closes #153 

**Steps necessary to review your pull request (required)**:
1. Checkout branch, `npm install`, `npm run start`
2. Go to http://localhost:4300/ids-data-grid/export-excel.html
3. Try both `Export to .xlsx` and `Export to .csv` buttons
4. Check that excel files successfully download
5. Check that excel files open without an error message
6. Check that datagrid data matches up with excel data

**Test special characters/locales**
1. The last 3 rows of data grid have special characters/different languages
2. Check that those special characters are rendered correctly in excel
3. Go to http://localhost:4300/ids-data-grid/example.html
4. Change locale to something like Bulgarian to change date/time formats
5. Run `$('ids-data-grid').exportToExcel('xlsx', 'test');` and `$('ids-data-grid').exportToExcel('csv', 'test');` in console
6. Check that the excel file date/times keep the format from datagrid in selected locale

Note: Columns with number/decimal/integer formats are passed as raw number values.
So, formatting from data grid for numeric type values aren't reflected in excel sheet.
This allows excel to parse those values as numeric values, rather than formatted strings, which allows 
excel user to perform their own excel calculations/formatting. It also prevents excel `numberStoredAsText` 
warnings to convert potential numeric string cells into numeric value cells.
This can be changed if requested.

**Test .xlsx zip**
1. Try extracting the .xlsx zip file. (I used 7zip -> extract to)
2. It should successfully unzip without error and reveal internal directories/files of .xlsx file
****

**Included in this Pull Request**:
